contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+        address _pool;
+        for (uint256 i = 0; i < _length; ++i) {
+            _pool = _validators[operatorAddresses[i]].creditContract;
+            require(_pool != address(0), "validator not exist");
+            stakeCredits[i] = _pool;
+        }
+
+        IGovToken(GOV_TOKEN_ADDR).sync(stakeCredits, account);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address consensusAddress) external payable onlyValidatorContract {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        Validator memory valInfo = _validators[operatorAddress];
+        if (valInfo.creditContract == address(0) || valInfo.jailed) {
+            emit RewardDistributeFailed(operatorAddress, "INVALID_VALIDATOR");
+            return;
+        }
+
+        try IStakeCredit(valInfo.creditContract).distributeReward{ value: msg.value }(valInfo.commission.rate) {
+            emit RewardDistributed(operatorAddress, msg.value);
+        } catch Error(string memory reason) {
+            emit RewardDistributeFailed(operatorAddress, bytes(reason));
+        } catch (bytes memory lowLevelData) {
+            emit RewardDistributeFailed(operatorAddress, lowLevelData);
+        }
+    }
+
+    /**
+     * @dev Get new eligible validators from consensus engine
+     */
+    function updateEligibleValidatorSet(
+        address[] calldata validators,
+        uint64[] calldata votingPowers,
+        bytes[] calldata voteAddrs
+    ) external onlyCoinbase onlyZeroGasPrice {
+        uint256 _length = validators.length;
+        if (_length == 0) {
+            return;
+        }
+
+        IBSCValidatorSet.Validator[] memory bscValidatorSet = new IBSCValidatorSet.Validator[](_length);
+        for (uint256 i; i < _length; ++i) {
+            bscValidatorSet[i] = IBSCValidatorSet.Validator({
+                consensusAddress: validators[i],
+                feeAddress: payable(address(0)),
+                BBCFeeAddress: address(0),
+                votingPower: votingPowers[i],
+                jailed: false,
+                incoming: 0
+            });
+        }
+
+        // update validator set in BSCValidatorSet contract
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).updateValidatorSetV2(bscValidatorSet, voteAddrs);
+    }
+
+    function downtimeSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(downtimeSlashAmount);
+        _jailValidator(valInfo);
+
+        uint256 jailUntil = block.timestamp + downtimeJailTime;
```

Comment:
can we move these logic inside the "_jailValidator(valInfo, jailUntil)"

Reply:
Done